codec-dns/src/main/java/io/netty/handler/codec/dns/AbstractDnsMessage.java

```diff
@@ -0,0 +1,344 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.dns;
+
+import io.netty.util.AbstractReferenceCounted;
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.ReferenceCounted;
+import io.netty.util.ResourceLeak;
+import io.netty.util.ResourceLeakDetector;
+import io.netty.util.internal.ObjectUtil;
+
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * A skeletal implementation of {@link DnsMessage}.
+ */
+public abstract class AbstractDnsMessage extends AbstractReferenceCounted implements DnsMessage {
+
+    private static final ResourceLeakDetector<DnsMessage> leakDetector =
+            new ResourceLeakDetector<DnsMessage>(DnsMessage.class);
+
+    private final ResourceLeak leak = leakDetector.open(this);
+    private short id;
+    private DnsOpCode opCode;
+    private boolean recursionDesired;
+    private byte z;
+
+    private Object questions; // a single question or the list of questions
+    private List<DnsRecord> answers;
+    private List<DnsRecord> authorityRecords;
+    private Object additionalRecords; // a single record or the list of records.
+
+    /**
+     * Creates a new instance with the specified {@code id} and {@link DnsOpCode#QUERY} opCode.
+     */
+    protected AbstractDnsMessage(int id) {
+        this(id, DnsOpCode.QUERY);
+    }
+
+    /**
+     * Creates a new instance with the specified {@code id} and {@code opCode}.
+     */
+    protected AbstractDnsMessage(int id, DnsOpCode opCode) {
+        setId(id);
+        setOpCode(opCode);
+    }
+
+    @Override
+    public int id() {
+        return id & 0xFFFF;
+    }
+
+    @Override
+    public DnsMessage setId(int id) {
+        this.id = (short) id;
+        return this;
+    }
+
+    @Override
+    public DnsOpCode opCode() {
+        return opCode;
+    }
+
+    @Override
+    public DnsMessage setOpCode(DnsOpCode opCode) {
+        ObjectUtil.checkNotNull(opCode, "opCode");
```

Comment:
static import checkNotNull?  Will make it easier to transition to java standard [requireNonNull](http://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#requireNonNull(T,%20java.lang.String)).  See issue https://github.com/netty/netty/issues/3500.

Global comment.